### PR TITLE
JavaScript: capture dynamically computed property names

### PR DESCRIPTION
**Pull request by masatake**
Signed-off-by: Masatake YAMATO <yamato@redhat.com>

### CODE REVIEW

**Comment by masatake** [comment]
@b4n, can I ask you to look at this one?
Your comment at https://github.com/universal-ctags/ctags/pull/1989 inspires me to write this change. 
Sorry for asking many things to you.

Input:
```JavaScript
var p = {
	[(1+2)*3]: {
		subnum: function (){}
	},
};
```
ctags with this change emits:
```
subnum	input.js	/^		subnum: function (){}$/;"	m	class:p.[(1+2)*3]
[(1+2)*3]	input.js	/^	[(1+2)*3]: {$/;"	c	class:p
p	input.js	/^var p = {$/;"	c
```

I wonder this is the way to go or not.
I have tried to write codes for ctags to capture as many things as possible.
My imaginably client tool utilizes the tags effectively.

However, I feel capturing dynamically computed property names is a bit too aggressive.

So I would like to ask you two questions before getting your review about code quality.

1. Does this change implement what you suggested in your comment in #1989 ?
2. (if the answer for 1. is yes) is this useful (in Geany e.g.) ?

One case this is really useful is this change may be able to capture `[Symbol.iterator]` as a generator in the following code:
```JavaScript
class InfiniteCounter {
  *[Symbol.iterator]() {
    for (let i = 0;; ++i) yield i
   }
}
```
(Suggested by  @maxnordlund  in #1949).

**Comment by maxnordlund** [comment]
Thank you for the great work 😄 

The most common use case for me at least is the [well-known symbols](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol#Well-known_symbols) like in the example I wrote.

Keeping the expression as-is and wrapped in brackets is intuitive as it matches what I have written.

File: parsers/jscript.c [hunk 1]
```diff
@@ -1717,6 +1732,12 @@ static bool parseMethods (tokenInfo *const token, const tokenInfo *const class,
 				{
 						bool has_child_methods = false;
 
+						if (is_dynamic_prop)
+						{
+							injectDynamicName (name, dprop);
+							dprop = NULL;
+						}
```
[the remark this answers is not in the document]
Comment (masatake) [reply parsers/jscript.c]: You are a great reviewer!

**Comment by masatake** [comment]
BTW, @maxnordlund, Coud you help us to improve JavaScript parser?

Could you recommend me some large JavaScript based open source software?
I'm building a codebase (https://github.com/universal-ctags/codebase) that is for evaluating parsers in ctags, especially about performance. I don't know JavaScript itself and its trend. So if you know source repositories in which not-small and well complicated JavaScript source files are, I would like to add them to our codebase. So we can detect performance degrading when adding new code to JavaScript parser of ctags. If you know good ones, please let me know them via a newly opend issue on https://github.com/universal-ctags/codebase .

File: parsers/jscript.c [hunk 2]
```diff
@@ -1667,6 +1667,13 @@ static bool parseMethods (tokenInfo *const token, const tokenInfo *const class,
 				} while (! isType (token, TOKEN_EOF) && depth > 0);
 			}
 
+			if (is_dynamic_prop)
+			{
+				injectDynamicName (name, dprop);
+				dprop = NULL;
+			}
+			vStringDelete (dprop);
```
Comment (b4n) [inline parsers/jscript.c]: the `vStringDelete (dprop)` could be in an else block (although it's safe as is as you set it to `NULL` in the if)

**Comment by masatake** [comment]
@b4n, I found one abnormal situation.

```
var p = {
	[(1+2)*3]: {
		subnum: function (){}
	},
	"[(1+2)*3]": {
		subnum2: function (){}
	}
};
```

In this case ctags emits two ""[(1+2)*3]" like
```
subnum	input.js	/^		subnum: function (){}$/;"	kind:method	line:42	language:JavaScript	scope:class:p.[(1+2)*3]	signature:()	roles:def
[(1+2)*3]	input.js	/^	[(1+2)*3]: {$/;"	kind:class	line:41	language:JavaScript	scope:class:p	roles:def
subnum2	input.js	/^		subnum2: function (){}$/;"	kind:method	line:45	language:JavaScript	scope:class:p.[(1+2)*3]	signature:()	roles:def
[(1+2)*3]	input.js	/^	"[(1+2)*3]": {$/;"	kind:class	line:44	language:JavaScript	scope:class:p	roles:def
```

Patterns after names are help users to distinguish two names.
I wonder it is enough or not. Basically I think ctags provides more hints for users to distinguish them if possible. Client tools can ignore them if they wants.

Two approaches; (1) is introducing a new kind, "dprop". For dynamic properties, assign "dprop" kind instead of "property", and (2) is attaching "dprop:" or "dynamic:"  JavaScript parser specific field.

Do you have comments for this?

File: parsers/jscript.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (masatake) [reply parsers/jscript.c]: Updated.

**Comment by masatake** [comment]
> ...
> Whether or not to include []s in the tag is questionable, but I kinda like it as it's also a visual cue that it's not a simple name.

Thank you for the comments. Yes,  questionable. Like full qualified tags, people who don't know ctags well may love the visual cue:-P.

"visual cue" is very important term. I'm looking for such words representing the concept in ctags.
Full qualifies tags are a kind of "visual cue". If a client tools are perfectly utilizing the information in tags file and u-ctags emits perfect tags file, "visual cue" is not needed. Synthesising  "visual cue" is a job of the client tool, not u-ctags. But we don't have perfects ones now. As far as not so redundant nor noisy,  preparing "visual cue" is not so bad thing.

File: parsers/jscript.c (hunk 2 above)
Comment (masatake) [reply parsers/jscript.c]: Updated.

**Comment by maxnordlund** [comment]
> Could you recommend me some large JavaScript based open source software?

@masatake done it, if anyone's curious [here they are](https://github.com/universal-ctags/codebase/pulls/maxnordlund).

**Comment by masatake** [comment]
@maxnordlund, thank you.
```
[jet@localhost]~/var/codebase% ./codebase ctags JavaScript maximum0 | grep -v Warn
version: 84dbaf52
features: +wildcards +regex +iconv +option-directory +xpath +json +interactive +sandbox +yaml +aspell
log: results/84dbaf52,JavaScript..........,maximum0..,time......,default...,2019-02-14-03:14:45.log
tagsoutput: /dev/null
cmdline: + u-ctags --quiet --options=NONE --sort=no --options=profile.d/maps --options=profile.d/maximum0.ctags --totals=yes --languages=JavaScript -o - -R code/acorn code/babel code/cockpit code/eslint code/katello code/pdf-js
10524 files, 884547 lines (27513 kB) scanned in 0.7 seconds (38422 kB/s)
88036 tags added to tag file

real	0m0.724s
user	0m0.605s
sys	0m0.117s
+ set +x
[jet@localhost]~/var/codebase% ./codebase ctags JavaScript minimum0 | grep -v Warn
version: 84dbaf52
features: +wildcards +regex +iconv +option-directory +xpath +json +interactive +sandbox +yaml +aspell
log: results/84dbaf52,JavaScript..........,minimum0..,time......,default...,2019-02-14-03:14:52.log
tagsoutput: /dev/null
cmdline: + u-ctags --quiet --options=NONE --sort=no --options=profile.d/maps --options=profile.d/minimum0.ctags --totals=yes --languages=JavaScript -o - -R code/acorn code/babel code/cockpit code/eslint code/katello code/pdf-js
10524 files, 884547 lines (27513 kB) scanned in 0.6 seconds (44369 kB/s)
65179 tags added to tag file

real	0m0.625s
user	0m0.507s
sys	0m0.116s
+ set +x
[jet@localhost]~/var/codebase% ./codebase ctags Go maximum0 | grep -v Warn        
version: 84dbaf52
features: +wildcards +regex +iconv +option-directory +xpath +json +interactive +sandbox +yaml +aspell
log: results/84dbaf52,Go..................,maximum0..,time......,default...,2019-02-14-03:15:05.log
tagsoutput: /dev/null
cmdline: + u-ctags --quiet --options=NONE --sort=no --options=profile.d/maps --options=profile.d/maximum0.ctags --totals=yes --languages=Go -o - -R code/buildah code/kubernetes
16397 files, 5116001 lines (162732 kB) scanned in 4.4 seconds (36935 kB/s)
1373368 tags added to tag file

real	0m4.423s
user	0m4.110s
sys	0m0.302s
+ set +x
[jet@localhost]~/var/codebase% ./codebase ctags Go minimum0 | grep -v Warn
version: 84dbaf52
features: +wildcards +regex +iconv +option-directory +xpath +json +interactive +sandbox +yaml +aspell
log: results/84dbaf52,Go..................,minimum0..,time......,default...,2019-02-14-03:15:17.log
tagsoutput: /dev/null
cmdline: + u-ctags --quiet --options=NONE --sort=no --options=profile.d/maps --options=profile.d/minimum0.ctags --totals=yes --languages=Go -o - -R code/buildah code/kubernetes
16397 files, 5116001 lines (162732 kB) scanned in 2.5 seconds (64107 kB/s)
627703 tags added to tag file

real	0m2.548s
user	0m2.315s
sys	0m0.226s
+ set +x
```

Here is the result on the PC.

**Comment by masatake** [comment]
@b4n, I merged this to avoid confliction between other queued changes.

However, 
```
var p = {
	[(1+2)*3]: {
		subnum: function (){}
	},
	"[(1+2)*3]": {
		subnum2: function (){}
	}
};
```
I would like to consider the above issue more.
I proposed two way to distinguish the properties: introducing new kind or attaching a parser specific field.
It seems that later one is better because the kind for `[(1+2)*3]` can be changed in the context, method,  setter, getter, generator, property, or class. Providing `dynamic` variants for each kind may not be a good idea. I will open a new PR for attaching the parser specific field. As far as reading `ctags-api.h`, there is no way to pass parser specific fields to geany from ctags. So attaching the `dprop` field (or `property:dynamically-computed`) has no impact to geany.

**Pull request merged**